File: stream.js
const webdriver = require('selenium-webdriver')
const chrome = require('selenium-webdriver/chrome')
const ytdl = require('ytdl-core')

class Video {
    async load(url, youtube_dl, msg) {
        if (this.in_loading) return
        this.in_loading = true
        this.driver.executeScript('video.innerHTML = null')

        if (youtube_dl) {
            await msg.edit("Fetching video formats...")
                .then(async msg => {
                    console.log("Fetching video formats...")
                    try {
                        let info = await ytdl.getInfo(url)
                        let formats = info.formats.filter(f => f.hasVideo && f.hasAudio)
                        formats = formats.filter(f => f.height <= 720 && f.fps <= 30)
                        formats = formats.sort((a, b) => b.height - a.height)

                        url = formats[0].url
                    }
                    catch (e) {
                        msg.edit(":no_entry_sign: " + String(e))
                    }
                })
        }

        await this.driver.executeScript(`video.src='${url}'`)
            .then(_ => {
                console.log('Loading...')
                msg.edit("Loading...")
                    .then(_ => {
                        var int1 = setInterval(() => {
                            is_error && clearInterval(int1)

                            if (this.killed) {
                                msg.edit(":no_entry_sign: Loading stopped")
                                this.in_loading = false
                                this.killed = false
                                clearInterval(int1)
                                clearInterval(int2)
                                clearInterval(int3)
                            }

                            this.driver.getCurrentUrl()
                                .then(url => {
                                    if (!this.init && url === "file:///channels/@me") {
                                        this.init = true
                                        this.open_guild()
                                        this.join(msg)
                                        clearInterval(int1)
                                    }
                                    else if (this.init)
                                        clearInterval(int1)
                                })
                        }, 10)
                    })
            })

        // Wait until video load
        let is_load
        var int2 = setInterval(() => {
            this.driver.executeScript("return video.duration")
                .then(result => {
                    if (result) {
                        is_load = true
                        this.duration = result
                        this.in_loading = false
                        msg.edit("Done, Type `*play` to start playing.")
                        clearInterval(int2)
                    }
                    else if (is_error)
                        clearInterval(int2)
                })
        }, 10)

        // Error event
        let is_error
        var int3 = setInterval(() => {
            this.driver.executeScript('return video_error')
                .then(error_msg => {
                    if (error_msg) {
                        msg.edit(":no_entry_sign: " + error_msg)
                        is_error = true
                        this.in_loading = false
                        this.driver.executeScript('video_error = ""')
                        clearInterval(int3)
                        return
                    }
                    else if (is_load)
                        clearInterval(int3)
                })
        }, 10)
    }

    play() {
        console.log("Play")
        this.start()
        this.driver.executeScript('video.play()')
    }

    pause() {
        console.log("Pause")
        this.driver.executeScript('video.pause()')
    }

    current(time = null) {
        if (time) {
            if (time[0] === '+' || time[0] === '-') {
                this.current().then(c => {
                    if (!c) return

                    let r
                    c = parseFloat(c)
                    const s = parseInt(time.slice(1))

                    time[0] === '+' ?
                        r = c + s :
                        r = c - s

                    this.driver.executeScript(`video.currentTime = ${r}`)
                })
            }
            else
                this.driver.executeScript(`video.currentTime = ${time}`)
        }
        else
            return this.driver.executeScript("return video.currentTime")
    }

    hms(sec) {
        if (sec)
            return new Date(sec * 1000).toISOString().substr(11, 8)
        return sec
    }
}

class Stream extends Video {
    client_url = `file://${__dirname}/client/index.html`

    constructor(token, headless = true) {
        super()
        const chrome_options = new chrome.Options()
        headless && chrome_options.addArguments('--headless')
        chrome_options.addArguments('--no-sandbox')
        chrome_options.addArguments('--window-size=1920,1080')
        chrome_options.addArguments('--disable-web-security')
        chrome_options.addArguments("--disable-gpu")
        chrome_options.addArguments("--disable-features=NetworkService")
        chrome_options.addArguments('--disable-dev-shm-usage')
        chrome_options.addArguments('--autoplay-policy=no-user-gesture-required')
        chrome_options.addArguments('user-agent=Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.50 Safari/537.36')
        console.log("Webdriver started")
        this.driver = new webdriver.Builder().forBrowser('chrome').setChromeOptions(chrome_options).build()
        this.driver.get(this.client_url)
        this.driver.executeScript(`localStorage.setItem("token", '"${token}"')`)
    }

    open_guild() {
        this.driver.executeScript(`document.querySelector('[data-list-item-id="guildsnav___${this.guild_id}"]').click()`)
    }

    is_full() {
        return this.driver.executeScript(`
            return document.querySelector("[aria-label='Channel is full']")
        `)
    }

    is_locked() {
        return this.driver.executeScript(`
            return document.querySelector("[data-list-item-id='channels___${this.channel_id}']").innerHTML.includes("Voice (Locked)")
        `)
    }

    scroll() {
        this.driver.executeScript(`
            var c_inject = document.getElementById("channels");
            if( c_inject.scrollTop === (c_inject.scrollHeight - c_inject.offsetHeight))
                c_inject.scroll(0, 0)
            else
                c_inject.scroll(0, c_inject.scrollTop + 250)
        `)
    }

    join(msg) {
        var intJoin = setInterval(() => {
            this.driver.executeScript(`document.querySelector("[data-list-item-id='channels___${this.channel_id}']").click()`)
                .then(() => {
                    // this.is_locked()
                    //     .then(result => {
                    //         if (result) {
                    //             msg.channel.send(":no_entry_sign: Channel is locked")
                    //             return
                    //         }
                    //     })

                    // this.is_full()
                    //     .then(result => {
                    //         if (result) {
                    //             msg.channel.send(":no_entry_sign: Channel is full")
                    //             return
                    //         }
                    //     })

                    setTimeout(() => {
                        this.start()
                    }, 1000)

                    clearInterval(intJoin)
                })
                .catch(() => this.scroll())
        }, 10)
    }

    start() {
        this.driver.executeScript(`
                var streamBtn_inject = document.querySelector('[aria-label="Share Your Screen"]')
                !streamBtn_inject.className.includes('buttonActive-3FrkXp') &&
                    streamBtn_inject.click()
        `).catch(e => e)
    }

    stop() {
        console.log("Stop")
        this.init = false
        this.driver.get(this.client_url)
    }
}

exports.Stream = Stream

File: stream.py
import undetected_chromedriver as uc
uc.install()

import os
import time
from selenium.webdriver import Chrome, ChromeOptions

class Video:
	def load_video(self, url):
		print("Loading video...")
		self.driver.execute_script(f"video.src='{url}'")
		
		# Wait until video load
		while not self.driver.execute_script("return video.readyState > 0"):
			time.sleep(0.05)

		self.duration = self.driver.execute_script("return video.duration")

	def play(self):
		print("Play")
		self.driver.execute_script("video.play()")

	def pause(self):
		print("Pause")
		self.driver.execute_script("video.pause()")

	def current(self, time=None):
		if time:
			self.driver.execute_script(f"video.currentTime = {time}")
		else:
			return self.driver.execute_script("return video.currentTime")

class Stream(Video):
	client_url = f"file://{os.getcwd()}/client/index.html"

	def __init__(self, guild_id, channel_id, headless=True):
		self.guild_id = guild_id
		self.channel_id = channel_id

		# Open index page
		options = ChromeOptions()
		options.add_argument('--headless') if headless else None
		options.add_argument('--no-sandbox')
		options.add_argument('--window-size=1920,1080')
		options.add_argument('--disable-web-security')
		options.add_argument('--disable-dev-shm-usage')
		options.add_argument('-–allow-file-access-from-files')
		options.add_argument('--autoplay-policy=no-user-gesture-required')
		self.driver = Chrome(executable_path="./chromedriver", options=options)
		print("Opening page...")
		self.driver.get(self.client_url)

	def open_guild(self):
		print("Opening guild...")
		while True:
			try:
				self.driver.execute_script(f'document.querySelector(\'[data-list-item-id="guildsnav___{self.guild_id}"]\').click()')
				break
			except:
				time.sleep(0.1)

	def is_full(self):
		if self.driver.execute_script(f'return document.querySelector("[aria-label=\'Channel is full\']")'):
			return True
		return False

	def is_locked(self):
		if self.driver.execute_script(f'return document.querySelector("[data-list-item-id=\'channels___{self.channel_id}\']").innerHTML.includes("Voice (Locked)")'):
			return True
		return False

	def scroll(self):
		self.driver.execute_script('''
			var c_inject = document.getElementById("channels");
			if( c_inject.scrollTop === (c_inject.scrollHeight - c_inject.offsetHeight))
				c_inject.scroll(0, 0)
			else
				c_inject.scroll(0, c_inject.scrollTop + 250)
		''')

	def join(self):
		print("Joining voice channel...")
		while True:
			try:
				self.scroll()

				if self.is_locked():
					print("Channel is locked")
					break

				self.driver.execute_script(f'document.querySelector("[data-list-item-id=\'channels___{self.channel_id}\']").click()')

				if self.is_full():	
					print("Channel is full")
					break

				print("Joined")
				break
			except:
				time.sleep(0.01)  

	def start(self):
		while True:
			try:
				self.driver.execute_script('document.querySelector(\'[aria-label="Share Your Screen"]\').click()')
				break
			except:
				time.sleep(0.1)

	def stop(self):
		self.driver.close()
		self.driver.quit()
